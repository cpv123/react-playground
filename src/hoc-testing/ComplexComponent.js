import React from 'react';
import PropTypes from 'prop-types';
import { Button, Typography, withStyles  } from '@material-ui/core';
import { compose, withState, withHandlers, onlyUpdateForKeys } from 'recompose';

const styles = {
    root: {
        display: 'flex',
        flexDirection: 'column',
        alignItems: 'center',
    },
};

export function ComplexComponent({ 
    classes, 
    heading, 
    incrementButtonTitle, 
    buttonTitle 
}) {
    return (
        <div className={classes.root}>
            <Typography>{heading}</Typography>
            <Button
                variant="contained"
                color="primary"
                onClick={incrementButtonTitle}
            >
                {buttonTitle}
            </Button>
        </div>
    );
}

ComplexComponent.propTypes = {
    classes: PropTypes.object.isRequired,
    heading: PropTypes.string.isRequired,
    incrementButtonTitle: PropTypes.func.isRequired,
    buttonTitle: PropTypes.number.isRequired,
};

const ComplexComponentWithHocs = compose(
    withStyles(styles),
    withState('buttonTitle', 'setButtonTitle', 0),
    withHandlers({
        incrementButtonTitle: ({ buttonTitle, setButtonTitle }) => () => {
            setButtonTitle(buttonTitle => buttonTitle + 1)
        }
    }),
    onlyUpdateForKeys([ "buttonTitle" ])
)(ComplexComponent);

export default ComplexComponentWithHocs;