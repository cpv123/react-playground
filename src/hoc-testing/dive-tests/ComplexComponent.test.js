import React from 'react';
import { shallow } from 'enzyme';
import ComplexComponent from '../ComplexComponent';

describe('The ComplexComponent component', () => {

    it('renders as expected', () => {
        const wrapper = shallow(
            <ComplexComponent heading="Test heading" /> // Provide the 'heading' prop only as the others are injected
        ).dive().dive().dive().dive();  // "one dive per HOC"
        
        expect(wrapper).toMatchSnapshot();
    });
    
});