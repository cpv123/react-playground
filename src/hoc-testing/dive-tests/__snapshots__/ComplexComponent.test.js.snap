// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The ComplexComponent component renders as expected 1`] = `
<div
  className="withState-withHandlers-onlyUpdateForKeys-ComplexComponent----root-1"
>
  <WithStyles(Typography)>
    Test heading
  </WithStyles(Typography)>
  <WithStyles(Button)
    color="primary"
    onClick={[Function]}
    variant="contained"
  >
    0
  </WithStyles(Button)>
</div>
`;
